Assignment 7

ThomasdenH opened the pull request:
My submission for Assignment 7, the AMM! https://github.com/yieldprotocol/mentorship2022/issues/7

alcueca commented on Assignment 7/src/Market.sol:
```diff
@@ -0,0 +1,218 @@
+    event SoldX(
+        address indexed account,
+        uint256 amount1,
+        uint256 amount2
```
These would be clearer if they would be `xAmount` and `yAmount`

alcueca commented on Assignment 7/src/Market.sol:
```diff
@@ -0,0 +1,218 @@
+    }
+
+    /// @notice Initialize this contract by supplying initial liquidity. This
+    ///     will also determine the intial price: `x` of token X and `y` of
```
typo: initial

alcueca commented on Assignment 7/src/Market.sol:
```diff
@@ -0,0 +1,218 @@
+
+    /// @notice Initialize this contract by supplying initial liquidity. This
+    ///     will also determine the intial price: `x` of token X and `y` of
+    ///     token Y should have equal value. This function can only be called
```
Not necessarily, you can initialize with any amount of x and y, the market will find their relative values.

alcueca commented on Assignment 7/src/Market.sol:
```diff
@@ -0,0 +1,218 @@
+                // Can be slightly smaller than z_y, due to rounding.
+                // Compute to ensure we don't give out too many liquidity tokens.
+                // We have z <= zY < zX
+                z = x * y0;
```
Isn't `z == zY`? Are you rounding on purpose?

ThomasdenH replied:
Yeah, that was my idea. Giving out slightly too many liquidity tokens could mean that you would be able to take out more underlying tokens than you would be entitled to. Doing it repeatedly might be an attack vector? However, if you think it's a non-issue I'd rather reuse `zY`.

alcueca commented on Assignment 7/src/Market.sol:
```diff
@@ -0,0 +1,218 @@
+                // Can be slightly smaller than z_x, due to rounding.
+                // Compute to ensure we don't give out too many liquidity tokens.
+                // We have z <= zX <= zY
+                z = y * x0;
```
`z = zX`?

alcueca commented on Assignment 7/src/Market.sol:
```diff
@@ -0,0 +1,218 @@
+
+        _mint(msg.sender, z);
+        tokenX.safeTransferFrom(msg.sender, address(this), x);
+        tokenY.safeTransferFrom(msg.sender, address(this), y);
```
You could take here only the tokens needed.

ThomasdenH replied:
I might not understand what you mean. I already recompute `x` or `y` so that - rounding aside - `z/z0 = x/x0 = y/y0`.

alcueca commented on Assignment 7/src/Market.sol:
```diff
@@ -0,0 +1,218 @@
+        // Here, we need
+        //      x0 * y0 = (x0 + x) * (y0 - y)
+        //      y = y0 - x0 * y0 / (x0 + x)
+        //        = y0 * x / (x0 + x)
```
Hold on, `y0 - x0 * y0 == y0 * x`?

ThomasdenH replied:
I've written this in a bit of a confusing way. I've updated it as follows:
```
y = y0 - (x0 * y0 / (x0 + x))
   = (y0 * x0 + y0 * x - x0 * y0) / (x0 + x)
   = y0 * x / (x0 + x)
```

alcueca commented on Assignment 7/test/Market.t.sol:
```diff
@@ -0,0 +1,236 @@
+        vm.stopPrank();
+
+        // Test that the market liquidity is enough to cover all liquidity
+        // tokens.
```
The invariant is that `k` is equal to the product of the reserves, and stays constant or increases with each trade. `k` is not equal to `totalSupply` except on initialization. `k != z`.

ThomasdenH replied:
Oof. I'll add more tests so that this is covered by them.

ThomasdenH commented:
I updated the contract and tests in accordance with the suggestions. I thought the invariant `k = x * y` was equal to `totalSupply()` and used that in the derivation of the maths. I therefore rewrote most of the calculations and derivation. I think it's much easier to understand now, if only for the fact that the calculations are much more straightforward.

As for tests, I added tests for events and errors, which I hadn't yet. Furthermore, I added tests that might have caught my wrong assumption by selling tokens after minting more liquidity (thereby breaking the assumption that `z=k`).

alcueca commented on Assignment 7/src/Market.sol:
```diff
@@ -0,0 +1,262 @@
+
+        _mint(msg.sender, z);
+
+        emit Initialized(msg.sender, x, y);
```
It's customary to emit events at the end of a function. `z` might be useful in the event as well.

alcueca commented on Assignment 7/src/Market.sol:
```diff
@@ -0,0 +1,262 @@
+    function initialize(uint256 x, uint256 y) external returns (uint256 z) {
+        require(_totalSupply == 0);
+
+        z = x * y;
```
If a user sends only `x` or `y` the effect will be a bit funny. The only impact is that an Initialized event will be emitted, without the contract actually being initialized. The user will lose some tokens, but that's from its own hand. I would probably put a `require` since you do it for `mint` as well.

alcueca reviewed (approved):
Two minor corrections, that don't subtract from the good work, really.

In terms of variable naming, I was discussing about a standard with @devtooligan this morning, we came up with:

Variable names that refer to contracts are in the form: `xyzToken`. Examples: xToken, yvToken, sharesToken.
In some rare cases an extra word will be needed: yvDaiToken if the contract also has a yvUsdcToken.
Token amounts will be referred as the contract variable, without `Token`, plus a word denoting what happens with the token. In and Out will be the preferred words to denote tokens being transferred into the contract, or out of the contract. Examples: xIn, yOut, sharesMinted, yvDaiReceived.
